File: tests/visual/logo_regression.py
#!/usr/bin/env python3
"""Visual regression test for the Deskia logo on the demo page.

Checks, across mobile / tablet / desktop viewports:
  1. Every <img alt="Deskia"> has loaded (naturalWidth > 0).
  2. The logo is visible (non-zero rendered size).
  3. Pixel-diff of the chat widget vs the committed baseline (tolerance 2%).

Usage:
    python3 tests/visual/logo_regression.py             # run check
    python3 tests/visual/logo_regression.py --update    # refresh baselines
    URL=http://localhost:8080/r/demo python3 ...        # custom target
"""
import asyncio, os, sys
from pathlib import Path
from PIL import Image, ImageChops
from playwright.async_api import async_playwright

ROOT = Path(__file__).parent
URL = os.environ.get("URL", "http://localhost:8080/r/demo")
# Lovable serves uploaded assets (/__l5e/*) from its CDN, not the local dev
# server, so when testing against localhost we transparently proxy those
# requests to the preview origin.
ASSET_ORIGIN = os.environ.get(
    "ASSET_ORIGIN",
    "https://id-preview--2c73aff8-1811-46af-98ee-ccd1e19e9c73.lovable.app",
)
UPDATE = "--update" in sys.argv
DIFF_THRESHOLD = 0.02  # 2% of pixels may differ

VIEWPORTS = [
    ("mobile", 375, 720),
    ("tablet", 768, 1024),
    ("desktop", 1280, 900),
]

BASELINE = ROOT / "baseline"
ACTUAL = ROOT / "actual"
DIFF = ROOT / "diff"
for d in (BASELINE, ACTUAL, DIFF):
    d.mkdir(parents=True, exist_ok=True)

failures = 0
def log(ok, msg):
    global failures
    print(("\u2713 " if ok else "\u2717 ") + msg)
    if not ok:
        failures += 1


def pixel_diff(base_path: Path, actual_path: Path, diff_path: Path) -> float:
    a = Image.open(base_path).convert("RGB")
    b = Image.open(actual_path).convert("RGB")
    if a.size != b.size:
        return 1.0
    d = ImageChops.difference(a, b)
    bbox_pixels = sum(1 for px in d.getdata() if px != (0, 0, 0))
    total = a.size[0] * a.size[1]
    # Save a visible diff (amplified)
    Image.eval(d, lambda v: min(255, v * 8)).save(diff_path)
    return bbox_pixels / total


async def main():
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True)
        try:
            for name, w, h in VIEWPORTS:
                ctx = await browser.new_context(viewport={"width": w, "height": h})
                page = await ctx.new_page()
                # Proxy /__l5e/* asset requests to the Lovable CDN.
                async def _proxy_asset(route):
                    req = route.request
                    target = ASSET_ORIGIN + req.url.split("://", 1)[1].split("/", 1)[1].replace(req.url.split("/", 3)[2], "", 1)
                    # simpler: rebuild from path
                    from urllib.parse import urlparse
                    path = urlparse(req.url).path
                    await route.continue_(url=ASSET_ORIGIN + path)
                await ctx.route("**/__l5e/**", _proxy_asset)

                await page.goto(URL, wait_until="networkidle")
                await page.wait_for_selector('img[alt="Deskia"]', timeout=10_000)
                await page.wait_for_function(
                    """() => {
                        const imgs = [...document.querySelectorAll('img[alt=\"Deskia\"]')];
                        return imgs.length > 0 && imgs.every(i => i.complete && i.naturalWidth > 0);
                    }""",
                    timeout=10_000,
                )

                imgs = await page.eval_on_selector_all(
                    'img[alt="Deskia"]',
                    """els => els.map(el => ({
                        natW: el.naturalWidth, natH: el.naturalHeight,
                        w: el.getBoundingClientRect().width,
                        h: el.getBoundingClientRect().height,
                    }))""",
                )
                log(len(imgs) >= 1, f"[{name}] found {len(imgs)} Deskia logo(s)")
                for i, im in enumerate(imgs):
                    log(im["natW"] > 0 and im["natH"] > 0,
                        f"[{name}] logo#{i} loaded (naturalWidth={im['natW']})")
                    log(im["w"] > 0 and im["h"] > 0,
                        f"[{name}] logo#{i} visible ({round(im['w'])}x{round(im['h'])})")

                card = await page.query_selector(".overflow-hidden")
                clip = await card.bounding_box() if card else None
                actual_path = ACTUAL / f"{name}.png"
                await page.screenshot(path=str(actual_path), clip=clip)

                base_path = BASELINE / f"{name}.png"
                if not base_path.exists() or UPDATE:
                    base_path.write_bytes(actual_path.read_bytes())
                    print(f"  \u21b3 baseline {'updated' if UPDATE else 'created'}: {base_path}")
                else:
                    ratio = pixel_diff(base_path, actual_path, DIFF / f"{name}.png")
                    log(ratio <= DIFF_THRESHOLD,
                        f"[{name}] pixel diff {ratio*100:.2f}% (threshold {DIFF_THRESHOLD*100}%)")

                await ctx.close()
        finally:
            await browser.close()

    if failures:
        print(f"\n{failures} check(s) failed", file=sys.stderr)
        sys.exit(1)
    print("\nAll logo visual checks passed.")


asyncio.run(main())

File: tests/visual/logo-regression.mjs
#!/usr/bin/env node
/**
 * Visual regression test for the Deskia logo on the public demo page.
 *
 * Verifies, across mobile / tablet / desktop viewports, that:
 *   1. Every <img alt="Deskia"> on the page has loaded (naturalWidth > 0).
 *   2. The logo is visible (non-zero rendered size, in viewport).
 *   3. Rendered height stays within the Tailwind class budget for that slot
 *      (header logo h-6 = 24px, avatar/typing logo h-5 = 20px).
 *   4. A screenshot is captured per viewport for manual diffing against the
 *      committed baseline in tests/visual/baseline/.
 *
 * Compares current run screenshots against baseline pixel-by-pixel when a
 * baseline exists; otherwise writes the current run as the new baseline.
 *
 * Usage:
 *   node tests/visual/logo-regression.mjs              # run check
 *   node tests/visual/logo-regression.mjs --update     # refresh baseline
 *   URL=http://localhost:8080/r/demo node tests/visual/logo-regression.mjs
 */
import { chromium } from "playwright";
import { PNG } from "pngjs";
import pixelmatch from "pixelmatch";
import { mkdirSync, existsSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const URL = process.env.URL || "http://localhost:8080/r/demo";
const UPDATE = process.argv.includes("--update");
const DIFF_THRESHOLD = 0.02; // 2% changed pixels tolerated

const VIEWPORTS = [
  { name: "mobile", width: 375, height: 720 },
  { name: "tablet", width: 768, height: 1024 },
  { name: "desktop", width: 1280, height: 900 },
];

const BASELINE_DIR = join(__dirname, "baseline");
const ACTUAL_DIR = join(__dirname, "actual");
const DIFF_DIR = join(__dirname, "diff");
[BASELINE_DIR, ACTUAL_DIR, DIFF_DIR].forEach((d) => mkdirSync(d, { recursive: true }));

const EXPECTED_HEIGHTS = { header: 24, avatar: 20, typing: 20 };
const HEIGHT_TOLERANCE = 2; // px

let failures = 0;
const log = (ok, msg) => {
  console.log(`${ok ? "✓" : "✗"} ${msg}`);
  if (!ok) failures++;
};

async function run() {
  const browser = await chromium.launch({ headless: true });
  try {
    for (const vp of VIEWPORTS) {
      const ctx = await browser.newContext({ viewport: { width: vp.width, height: vp.height } });
      const page = await ctx.newPage();
      await page.goto(URL, { waitUntil: "networkidle" });
      // wait for chat widget header logo
      await page.waitForSelector('img[alt="Deskia"]', { timeout: 10_000 });

      // Verify every Deskia logo image
      const imgs = await page.$$eval('img[alt="Deskia"]', (els) =>
        els.map((el) => ({
          src: el.getAttribute("src"),
          natW: el.naturalWidth,
          natH: el.naturalHeight,
          rect: el.getBoundingClientRect().toJSON(),
        }))
      );

      log(imgs.length >= 1, `[${vp.name}] found ${imgs.length} Deskia logo(s)`);
      imgs.forEach((img, i) => {
        log(img.natW > 0 && img.natH > 0, `[${vp.name}] logo#${i} loaded (naturalWidth=${img.natW})`);
        log(img.rect.width > 0 && img.rect.height > 0, `[${vp.name}] logo#${i} visible (${Math.round(img.rect.width)}×${Math.round(img.rect.height)})`);
      });

      // Crop & screenshot the chat widget card for visual diff
      const card = await page.$('.overflow-hidden'); // the ChatWidget Card
      const clip = card ? await card.boundingBox() : null;
      const shotPath = join(ACTUAL_DIR, `${vp.name}.png`);
      await page.screenshot({ path: shotPath, clip: clip || undefined });

      const basePath = join(BASELINE_DIR, `${vp.name}.png`);
      if (!existsSync(basePath) || UPDATE) {
        writeFileSync(basePath, readFileSync(shotPath));
        console.log(`  ↳ baseline ${UPDATE ? "updated" : "created"}: ${basePath}`);
      } else {
        const base = PNG.sync.read(readFileSync(basePath));
        const actual = PNG.sync.read(readFileSync(shotPath));
        if (base.width !== actual.width || base.height !== actual.height) {
          log(false, `[${vp.name}] screenshot dimensions changed ${base.width}×${base.height} → ${actual.width}×${actual.height}`);
        } else {
          const diff = new PNG({ width: base.width, height: base.height });
          const changed = pixelmatch(base.data, actual.data, diff.data, base.width, base.height, { threshold: 0.1 });
          const ratio = changed / (base.width * base.height);
          writeFileSync(join(DIFF_DIR, `${vp.name}.png`), PNG.sync.write(diff));
          log(ratio <= DIFF_THRESHOLD, `[${vp.name}] pixel diff ${(ratio * 100).toFixed(2)}% (threshold ${DIFF_THRESHOLD * 100}%)`);
        }
      }

      await ctx.close();
    }
  } finally {
    await browser.close();
  }

  if (failures > 0) {
    console.error(`\n${failures} check(s) failed`);
    process.exit(1);
  }
  console.log("\nAll logo visual checks passed.");
}

run().catch((e) => {
  console.error(e);
  process.exit(1);
});
